packages/eip-7702-internal-rpc-middleware/src/wallet_getAccountUpgradeStatus.ts

```diff
@@ -0,0 +1,142 @@
+import { JsonRpcError, rpcErrors } from '@metamask/rpc-errors';
+import { tuple } from '@metamask/superstruct';
+import {
+  type JsonRpcRequest,
+  type PendingJsonRpcResponse,
+  type Json,
+  type Hex,
+  getErrorMessage,
+} from '@metamask/utils';
+
+import { DELEGATION_INDICATOR_PREFIX } from './constants';
+import { GetAccountUpgradeStatusParamsStruct } from './types';
+import { validateParams, validateAndNormalizeAddress } from './utils';
+
+export type WalletGetAccountUpgradeStatusHooks = {
+  getCurrentChainIdForDomain: (origin: string) => Hex | null;
+  getCode: (address: string, networkClientId: string) => Promise<string | null>;
+  getSelectedNetworkClientIdForChain: (chainId: string) => string | null;
+  getPermittedAccountsForOrigin: (origin: string) => Promise<string[]>;
+  isEip7702Supported: (request: { address: string; chainId: Hex }) => Promise<{
+    isSupported: boolean;
+    upgradeContractAddress?: string;
+  }>;
+};
+
+const isAccountUpgraded = async (
+  address: string,
+  networkClientId: string,
+  getCode: (address: string, networkClientId: string) => Promise<string | null>,
+): Promise<{ isUpgraded: boolean; upgradedAddress: Hex | null }> => {
+  const code = await getCode(address, networkClientId);
+  if (!code || code === '0x' || code.length <= 2) {
+    return { isUpgraded: false, upgradedAddress: null };
+  }
+
+  if (!code.startsWith(DELEGATION_INDICATOR_PREFIX)) {
+    return { isUpgraded: false, upgradedAddress: null };
+  }
+
+  const expectedLength = DELEGATION_INDICATOR_PREFIX.length + 40; // 0xef0100 + 40 hex chars
+  if (code.length !== expectedLength) {
+    return { isUpgraded: false, upgradedAddress: null };
+  }
+
+  // Extract the 20-byte address (40 hex characters after the prefix)
+  const upgradedAddress = `0x${code.slice(8, 48)}` as Hex;
+
+  return { isUpgraded: true, upgradedAddress };
+};
+
+/**
+ * The RPC method handler middleware for `wallet_getAccountUpgradeStatus`
+ *
+ * @param req - The JSON RPC request's end callback.
+ * @param res - The JSON RPC request's pending response object.
+ * @param hooks - The hooks required for account upgrade status checking.
+ */
+export async function walletGetAccountUpgradeStatus(
+  req: JsonRpcRequest<Json[]> & { origin: string },
```

Comment:
just wondering why we using array params. Not sure if this was discussed somewhere already or not